jetty-core/jetty-http3/jetty-http3-client/src/main/java/org/eclipse/jetty/http3/client/HTTP3SessionClient.java

```diff
@@ -91,49 +96,53 @@ public void onSettings(SettingsFrame frame)
     }
 
     @Override
-    public CompletableFuture<Stream> newRequest(HeadersFrame frame, Stream.Client.Listener listener)
+    public void newRequest(HeadersFrame frame, Stream.Client.Listener listener, Promise.Invocable<Stream> promise)
     {
-        long streamId = getProtocolSession().newStreamId(StreamType.CLIENT_BIDIRECTIONAL);
-        return newRequest(streamId, frame, listener);
-    }
+        var quicSession = getProtocolSession().getSession();
+        long streamId = quicSession.newStreamId(true);
+        var quicStream = quicSession.newStream(streamId, null);
 
-    private CompletableFuture<Stream> newRequest(long streamId, HeadersFrame frame, Stream.Client.Listener listener)
-    {
         if (LOG.isDebugEnabled())
-            LOG.debug("new request stream #{} with {} on {}", streamId, frame, this);
+            LOG.debug("new request {} with {} on {}", quicStream, frame, this);
 
         ProtocolSession session = getProtocolSession();
-        QuicStreamEndPoint endPoint = session.getOrCreateStreamEndPoint(streamId, session::openProtocolEndPoint);
+        StreamEndPoint endPoint = session.createStreamEndPoint(quicStream, session::openStreamEndPoint);
+        ((AbstractStream)quicStream).setListener(new ProtocolStreamListener(() -> endPoint));
```

Comment:
setting the listener a few lines after the construction looks sub-optimal.  OK, the stream needs to know the endPoint and the endPoint needs to know the stream, but that feels like there should be a setter for endPoint rather than for the listener.